File: objc/test/Ice/operations/TestI.m
// **********************************************************************
//
// Copyright (c) 2003-2008 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#import <Ice/Ice.h>

#import <TestI.h>
#import <TestCommon.h>

@implementation MyDerivedClassI
-(void) opVoid:(ICECurrent*)current
{
}

-(void) opDerived:(ICECurrent*)current
{
}

-(ICEByte) opByte:(ICEByte)p1 p2:(ICEByte)p2 p3:(ICEByte *)p3 current:(ICECurrent *)current
{
    *p3 = p1 ^ p2;
    return p1;
}

-(BOOL) opBool:(BOOL)p1 p2:(BOOL)p2 p3:(BOOL *)p3 current:(ICECurrent*)current
{
    *p3 = p1;
    return p2;
}

-(ICELong) opShortIntLong:(ICEShort)p1 p2:(ICEInt)p2 p3:(ICELong)p3
                          p4:(ICEShort *)p4 p5:(ICEInt *)p5 p6:(ICELong *)p6
		          current:(ICECurrent *)current
{
    *p4 = p1;
    *p5 = p2;
    *p6 = p3;
    return p3;
}

-(ICEDouble) opFloatDouble:(ICEFloat)p1 p2:(ICEDouble)p2 p3:(ICEFloat *)p3 p4:(ICEDouble *)p4
                           current:(ICECurrent *)current
{
    *p3 = p1;
    *p4 = p2;
    return p2;
}

-(NSString *) opString:(NSMutableString *)p1 p2:(NSMutableString *)p2 p3:(NSString **)p3 current:(ICECurrent *)current
{
    NSMutableString *sout = [NSMutableString stringWithCapacity:([p2 length] + 1 + [p1 length])];
    [sout appendString:p2];
    [sout appendString:@" "];
    [sout appendString:p1];
    *p3 = sout;

    NSMutableString *ret = [NSMutableString stringWithCapacity:([p1 length] + 1 + [p2 length])];
    [ret appendString:p1];
    [ret appendString:@" "];
    [ret appendString:p2];
    return ret;
}

-(TestMyEnum) opMyEnum:(TestMyEnum)p1 p2:(TestMyEnum *)p2 current:(ICECurrent *)current
{
    *p2 = p1;
    return Testenum3;
}

-(id<TestMyClassPrx>) opMyClass:(id<TestMyClassPrx>)p1 p2:(id<TestMyClassPrx> *)p2 p3:(id<TestMyClassPrx> *)p3 
                        current:(ICECurrent *)current
{
    *p2 = p1;
    *p3 = [TestMyClassPrx uncheckedCast:[[current adapter] 
                                           createProxy:[[[current adapter] getCommunicator] 
                                                           stringToIdentity:@"noSuchIdentity"]]];
    return [TestMyClassPrx uncheckedCast:[[current adapter] createProxy:[current id_]]];
}

-(TestStructure *) opStruct:(TestStructure *)p1 p2:(TestStructure *)p2 p3:(TestStructure **)p3
                            current:(ICECurrent *)current;
{
    *p3 = [[p1 copy] autorelease];
    (*p3).s.s = @"a new string";
    return p2;
}

-(TestByteS *) opByteS:(TestMutableByteS *)p1 p2:(TestMutableByteS *)p2 p3:(TestByteS **)p3
                       current:(ICECurrent *)current
{
    *p3 = [TestMutableByteS dataWithLength:[p1 length]];
    ICEByte *target = (ICEByte *)[*p3 bytes];
    ICEByte *src = (ICEByte *)[p1 bytes] + [p1 length];
    int i;
    for(i = 0; i != [p1 length]; ++i)
    {
        *target++ = *--src;
    }
    TestMutableByteS *r = [TestMutableByteS dataWithData:p1];
    [r appendData:p2];
    return r;
}

-(TestBoolS *) opBoolS:(TestMutableBoolS *)p1 p2:(TestMutableBoolS *)p2 p3:(TestBoolS **)p3
                       current:(ICECurrent *)current
{
    *p3 = [TestMutableBoolS dataWithData:p1];
    [(TestMutableBoolS *)*p3 appendData:p2];

    TestMutableBoolS *r = [TestMutableBoolS dataWithLength:[p1 length] * sizeof(BOOL)];
    BOOL *target = (BOOL *)[r bytes];
    BOOL *src = (BOOL *)([p1 bytes] + [p1 length]);
    int i;
    for(i = 0; i != [p1 length]; ++i)
    {
        *target++ = *--src;
    }
    return r;
}

-(TestLongS *) opShortIntLongS:(TestMutableShortS *)p1 p2:(TestMutableIntS *)p2 p3:(TestMutableLongS *)p3
                               p4:(TestShortS **)p4 p5:(TestIntS **)p5 p6:(TestLongS **)p6
			       current:(ICECurrent *)current
{
    *p4 = [TestMutableShortS dataWithData:p1];
    *p5 = [TestMutableIntS dataWithLength:[p2 length]];
    ICEInt *target = (ICEInt *)[*p5 bytes];
    ICEInt *src = (ICEInt *)([p2 bytes] + [p2 length]);
    int i;
    for(i = 0; i != [p2 length] / sizeof(ICEInt); ++i)
    {
        *target++ = *--src;
    }
    *p6 = [TestMutableLongS dataWithData:p3];
    [(TestMutableLongS *)*p6 appendData:p3];
    return p3;
}

-(TestDoubleS *) opFloatDoubleS:(TestMutableFloatS *)p1 p2:(TestMutableDoubleS *)p2
                                p3:(TestFloatS **)p3 p4:(TestDoubleS **)p4 current:(ICECurrent *)current
{
    *p3 = [TestMutableFloatS dataWithData:p1];
    *p4 = [TestMutableDoubleS dataWithLength:[p2 length]];
    ICEDouble *target = (ICEDouble *)[*p4 bytes];
    ICEDouble *src = (ICEDouble *)([p2 bytes] + [p2 length]);
    int i;
    for(i = 0; i != [p2 length] / sizeof(ICEDouble); ++i)
    {
        *target++ = *--src;
    }
    TestDoubleS *r = [TestMutableDoubleS dataWithLength:([p2 length]
                                                         + ([p1 length] / sizeof(ICEFloat) * sizeof(ICEDouble)))];
    ICEDouble *rp = (ICEDouble *)[r bytes];
    ICEDouble *p2p = (ICEDouble *)[p2 bytes];
    for(i = 0; i < [p2 length] / sizeof(ICEDouble); ++i)
    {
        *rp++ = *p2p++;
    }
    ICEFloat *bp1 = (ICEFloat *)[p1 bytes];
    for(i = 0; i < [p1 length] / sizeof(ICEFloat); ++i)
    {
        *rp++ = bp1[i];
    }
    return r;
}

-(TestStringS *) opStringS:(TestMutableStringS *)p1 p2:(TestMutableStringS *)p2
                           p3:(TestStringS **)p3 current:(ICECurrent *)current
{
    *p3 = [TestMutableStringS arrayWithArray:p1];
    [(TestMutableStringS *)*p3 addObjectsFromArray:p2];
    TestMutableStringS *r = [TestMutableStringS arrayWithCapacity:[p1 count]];
    NSEnumerator *enumerator = [p1 reverseObjectEnumerator];
    for(NSString *element in enumerator)
    {
        [r addObject:element];
    } 
    return r;
}

-(TestMyEnumS *) opMyEnumS:(TestMutableMyEnumS *)p1 p2:(TestMutableMyEnumS *)p2
                           p3:(TestMyEnumS **)p3 current:(ICECurrent *)current
{
    *p3 = [TestMutableMyEnumS dataWithLength:[p1 length]];
    TestMyEnum *target = (TestMyEnum *)[*p3 bytes];
    TestMyEnum *src = (TestMyEnum *)([p1 bytes] + [p1 length]);
    int i;
    for(i = 0; i != [p1 length] / sizeof(TestMyEnum); ++i)
    {
        *target++ = *--src;
    }
    TestMutableMyEnumS *r = [TestMutableMyEnumS dataWithData:p1];
    [r appendData:p2];
    return r;
}

-(TestMyClassS *) opMyClassS:(TestMutableMyClassS *)p1 p2:(TestMutableMyClassS *)p2
                          p3:(TestMyClassS **)p3 current:(ICECurrent *)current
{
    *p3 = [TestMutableMyClassS arrayWithArray:p1];
    [(TestMutableMyClassS *)*p3 addObjectsFromArray:p2];
    TestMutableMyClassS *r = [TestMutableMyClassS arrayWithCapacity:[p1 count]];
    NSEnumerator *enumerator = [p1 reverseObjectEnumerator];
    for(NSString *element in enumerator)
    {
        [r addObject:element];
    } 
    return r;
}

-(TestByteSS *) opByteSS:(TestMutableByteSS *)p1 p2:(TestMutableByteSS *)p2 p3:(TestByteSS * *)p3
                         current:(ICECurrent *)current
{
    *p3 = [TestMutableByteSS array];
    NSEnumerator *enumerator = [p1 reverseObjectEnumerator]; 
    for(TestByteS *element in enumerator)
    { 
        [(TestMutableByteSS *)*p3 addObject:element];
    }

    TestMutableByteSS *r = [TestMutableByteSS arrayWithArray:p1];
    [r addObjectsFromArray:p2];
    return r;
}

-(TestBoolSS *) opBoolSS:(TestMutableBoolSS *)p1 p2:(TestMutableBoolSS *)p2 p3:(TestBoolSS * *)p3
                         current:(ICECurrent *)current
{
    *p3 = [TestMutableBoolSS arrayWithArray:p1];
    [(TestMutableBoolSS *)*p3 addObjectsFromArray:p2];

    TestMutableBoolSS *r = [TestMutableBoolSS array];
    NSEnumerator *enumerator = [p1 reverseObjectEnumerator]; 
    for(TestBoolS *element in enumerator)
    { 
        [r addObject:element];
    }
    return r;
}

-(TestLongSS *) opShortIntLongSS:(TestMutableShortSS *)p1 p2:(TestMutableIntSS *)p2 p3:(TestMutableLongSS *)p3
                                 p4:(TestShortSS **)p4 p5:(TestIntSS **)p5 p6:(TestLongSS **)p6
			         current:(ICECurrent *)current
{
    *p4 = [TestShortSS arrayWithArray:p1];
    *p5 = [TestMutableIntSS array];
    NSEnumerator *enumerator = [p2 reverseObjectEnumerator]; 
    for(TestIntS *element in enumerator)
    { 
        [(TestMutableIntSS *)*p5 addObject:element];
    }
    *p6 = [TestMutableLongSS arrayWithArray:p3];
    [(TestMutableLongSS *)*p6 addObjectsFromArray:p3];
    return p3;
}

-(TestDoubleSS *) opFloatDoubleSS:(TestMutableFloatSS *)p1 p2:(TestMutableDoubleSS *)p2
                                  p3:(TestFloatSS **)p3 p4:(TestDoubleSS **)p4 current:(ICECurrent *)current
{
    *p3 = [TestFloatSS arrayWithArray:p1];
    *p4 = [TestMutableDoubleSS array];
    NSEnumerator *enumerator = [p2 reverseObjectEnumerator]; 
    for(TestDoubleS *element in enumerator)
    { 
        [(TestMutableDoubleSS *)*p4 addObject:element];
    }
    TestMutableDoubleSS *r = [TestMutableDoubleSS arrayWithArray:p2];
    [r addObjectsFromArray:p2];
    return r;
}

-(TestStringSS *) opStringSS:(TestMutableStringSS *)p1 p2:(TestMutableStringSS *)p2 p3:(TestStringSS **)p3
                             current:(ICECurrent *)current
{
    *p3 = [TestMutableStringSS arrayWithArray:p1];
    [(TestMutableStringSS *)*p3 addObjectsFromArray:p2];
    TestMutableStringSS *r = [TestMutableStringSS array];
    NSEnumerator *enumerator = [p2 reverseObjectEnumerator]; 
    for(TestStringS *element in enumerator)
    { 
        [r addObject:element];
    }
    return r;
}

-(TestStringSSS *) opStringSSS:(TestMutableStringSSS *)p1 p2:(TestMutableStringSSS *)p2 p3:(TestStringSSS **)p3
                               current:(ICECurrent *)current
{
    *p3 = [TestMutableStringSSS arrayWithArray:p1];
    [(TestMutableStringSSS *)*p3 addObjectsFromArray:p2];
    TestMutableStringSSS *r = [TestMutableStringSSS array];
    NSEnumerator *enumerator = [p2 reverseObjectEnumerator];
    for(TestStringSS *element in enumerator)
    { 
        [r addObject:element];
    }
    return r;
}

-(TestByteBoolD *) opByteBoolD:(TestMutableByteBoolD *)p1 p2:(TestMutableByteBoolD *)p2 p3:(TestByteBoolD **)p3
                               current:(ICECurrent *)current
{
    *p3 = [TestMutableByteBoolD dictionaryWithDictionary:p1];
    TestMutableByteBoolD *r = [TestMutableByteBoolD dictionaryWithDictionary:p1];
    [r addEntriesFromDictionary:p2];
    return r;
}

-(TestShortIntD *) opShortIntD:(TestMutableShortIntD *)p1 p2:(TestMutableShortIntD *)p2 p3:(TestShortIntD **)p3
                               current:(ICECurrent *)current
{
    *p3 = [TestMutableShortIntD dictionaryWithDictionary:p1];
    TestMutableShortIntD *r = [TestMutableShortIntD dictionaryWithDictionary:p1];
    [r addEntriesFromDictionary:p2];
    return r;
}

-(TestLongFloatD *) opLongFloatD:(TestMutableLongFloatD *)p1 p2:(TestMutableLongFloatD *)p2 p3:(TestLongFloatD **)p3
                               current:(ICECurrent *)current
{
    *p3 = [TestMutableLongFloatD dictionaryWithDictionary:p1];
    TestMutableLongFloatD *r = [TestMutableLongFloatD dictionaryWithDictionary:p1];
    [r addEntriesFromDictionary:p2];
    return r;
}

-(TestStringStringD *) opStringStringD:(TestMutableStringStringD *)p1 p2:(TestMutableStringStringD *)p2
                                       p3:(TestStringStringD **)p3 current:(ICECurrent *)current
{
    *p3 = [TestMutableStringStringD dictionaryWithDictionary:p1];
    TestMutableStringStringD *r = [TestMutableStringStringD dictionaryWithDictionary:p1];
    [r addEntriesFromDictionary:p2];
    return r;
}

-(TestStringMyEnumD *) opStringMyEnumD:(TestMutableStringMyEnumD *)p1 p2:(TestMutableStringMyEnumD *)p2
                                       p3:(TestStringMyEnumD **)p3 current:(ICECurrent *)current
{
    *p3 = [TestMutableStringMyEnumD dictionaryWithDictionary:p1];
    TestMutableStringMyEnumD *r = [TestMutableStringMyEnumD dictionaryWithDictionary:p1];
    [r addEntriesFromDictionary:p2];
    return r;
}

-(TestIntS *) opIntS:(TestMutableIntS *)p1 current:(ICECurrent *)current
{
    int count = [p1 length] / sizeof(ICEInt);
    TestMutableIntS *r = [TestMutableIntS dataWithLength:[p1 length]];
    const int *src = [p1 bytes];
    int *target = (int *)[r bytes];
    while(count-- > 0)
    {
	*target++ = -*src++;
    }
    return r;
}

-(void) opByteSOneway:(TestMutableByteS *)p1 current:(ICECurrent *)current
{
}

-(ICEContext *) opContext:(ICECurrent *)current
{
    return current.ctx;
}

-(void) opDoubleMarshaling:(ICEDouble)p1 p2:(TestMutableDoubleS *)p2 current:(ICECurrent *)current
{
    ICEDouble d = 1278312346.0 / 13.0;
    test(p1 == d);
    const ICEDouble *p = [p2 bytes];
    int i;
    for(i = 0; i < [p2 length] / sizeof(ICEDouble); ++i)
    {
	test(p[i] == d);
    }
}

-(TestStringS *) getNSNullStringSeq:(ICECurrent *)current
{
    return [NSArray arrayWithObjects:@"first", [NSNull null], nil];
}

-(TestMyClassS *) getNSNullASeq:(ICECurrent *)current
{
    return [NSArray arrayWithObjects:[TestA a:99], [NSNull null], nil];
}

-(TestStructS *) getNSNullStructSeq:(ICECurrent *)current
{
    TestStructure *s = [TestStructure structure:nil e:Testenum2 s:[TestAnotherStruct anotherStruct:@"Hello"]];
    return [NSArray arrayWithObjects:s, [NSNull null], nil];
}

-(TestStringSS *) getNSNullStringSeqSeq:(ICECurrent *)current
{
    TestStringSS *s = [NSArray arrayWithObjects:@"first", nil];
    return [NSArray arrayWithObjects:s, [NSNull null], nil];
}

-(TestStringStringD *) getNSNullStringStringDict:(ICECurrent *)current
{
    TestMutableStringStringD *d = [TestMutableStringStringD dictionary];
    [d setObject:@"ONE" forKey:@"one"];
    [d setObject:[NSNull null] forKey:@"two"];
    return d;
}

-(void) putNSNullStringStringDict:(TestMutableStringStringD *)d current:(ICECurrent *)current
{
    // Nothing to do because this tests that an exception is thrown on the client side.
}

-(void) putNSNullShortIntDict:(TestMutableShortIntD *)d current:(ICECurrent *)current
{
    // Nothing to do because this tests that an exception is thrown on the client side.
}

-(void) putNSNullStringMyEnumDict:(TestMutableStringMyEnumD *)d current:(ICECurrent *)current
{
    // Nothing to do because this tests that an exception is thrown on the client side.
}

-(void) shutdown:(ICECurrent*)current
{
    [[[current adapter] getCommunicator] shutdown];
}
@end

File: objc/test/cpp/Ice/operations/TestI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2011 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_TOUCH_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Ice/Ice.h>
#include <IceUtil/IceUtil.h>
#include <operations/TestI.h>
#include <TestCommon.h>
#include <functional>
#include <iterator>

using namespace Test::Operations;

bool
MyDerivedClassI::ice_isA(const std::string& id, const Ice::Current& current) const
{
    test(current.mode == Ice::Nonmutating);
    return Test::Operations::MyDerivedClass::ice_isA(id, current);
}

void
MyDerivedClassI::ice_ping(const Ice::Current& current) const
{
    test(current.mode == Ice::Nonmutating);
    Test::Operations::MyDerivedClass::ice_ping(current);
}

std::vector<std::string>
MyDerivedClassI::ice_ids(const Ice::Current& current) const
{
    test(current.mode == Ice::Nonmutating);
    return Test::Operations::MyDerivedClass::ice_ids(current);
}

const std::string&
MyDerivedClassI::ice_id(const Ice::Current& current) const
{
    test(current.mode == Ice::Nonmutating);
    return Test::Operations::MyDerivedClass::ice_id(current);
}

void
MyDerivedClassI::shutdown(const Ice::Current& current)
{
    current.adapter->getCommunicator()->shutdown();
}

void
MyDerivedClassI::delay(Ice::Int ms, const Ice::Current& current)
{
    IceUtil::ThreadControl::sleep(IceUtil::Time::milliSeconds(ms));
}

void
MyDerivedClassI::opVoid(const Ice::Current& current)
{
    test(current.mode == Ice::Normal);
}

Ice::Byte
MyDerivedClassI::opByte(Ice::Byte p1,
                        Ice::Byte p2,
                        Ice::Byte& p3,
                        const Ice::Current&)
{
    p3 = p1 ^ p2;
    return p1;
}

bool
MyDerivedClassI::opBool(bool p1,
                        bool p2,
                        bool& p3,
                        const Ice::Current&)
{
    p3 = p1;
    return p2;
}

Ice::Long
MyDerivedClassI::opShortIntLong(Ice::Short p1,
                                Ice::Int p2,
                                Ice::Long p3,
                                Ice::Short& p4,
                                Ice::Int& p5,
                                Ice::Long& p6,
                                const Ice::Current&)
{
    p4 = p1;
    p5 = p2;
    p6 = p3;
    return p3;
}

Ice::Double
MyDerivedClassI::opFloatDouble(Ice::Float p1,
                               Ice::Double p2,
                               Ice::Float& p3,
                               Ice::Double& p4,
                               const Ice::Current&)
{
    p3 = p1;
    p4 = p2;
    return p2;
}

std::string
MyDerivedClassI::opString(const std::string& p1,
                          const std::string& p2,
                          std::string& p3,
                          const Ice::Current&)
{
    p3 = p2 + " " + p1;
    return p1 + " " + p2;
}

Test::Operations::MyEnum
MyDerivedClassI::opMyEnum(Test::Operations::MyEnum p1,
                          Test::Operations::MyEnum& p2,
                          const Ice::Current&)
{
    p2 = p1;
    return Test::Operations::enum3;
}

Test::Operations::MyClassPrx
MyDerivedClassI::opMyClass(const Test::Operations::MyClassPrx& p1,
                           Test::Operations::MyClassPrx& p2,
                           Test::Operations::MyClassPrx& p3,
                           const Ice::Current& current)
{
    p2 = p1;
    p3 = Test::Operations::MyClassPrx::uncheckedCast(current.adapter->createProxy(
                                                                      current.adapter->getCommunicator()->stringToIdentity("noSuchIdentity")));
    return Test::Operations::MyClassPrx::uncheckedCast(current.adapter->createProxy(current.id));
}

Test::Operations::Structure
MyDerivedClassI::opStruct(const Test::Operations::Structure& p1,
                          const ::Test::Operations::Structure& p2,
                          ::Test::Operations::Structure& p3,
                          const Ice::Current&)
{
    p3 = p1;
    p3.s.s = "a new string";
    return p2;
}

Test::Operations::ByteS
MyDerivedClassI::opByteS(const Test::Operations::ByteS& p1,
                         const Test::Operations::ByteS& p2,
                         Test::Operations::ByteS& p3,
                         const Ice::Current&)
{
    p3.resize(p1.size());
    std::reverse_copy(p1.begin(), p1.end(), p3.begin());
    Test::Operations::ByteS r = p1;
    std::copy(p2.begin(), p2.end(), std::back_inserter(r));
    return r;
}

Test::Operations::BoolS
MyDerivedClassI::opBoolS(const Test::Operations::BoolS& p1,
                         const Test::Operations::BoolS& p2,
                         Test::Operations::BoolS& p3,
                         const Ice::Current&)
{
    p3 = p1;
    std::copy(p2.begin(), p2.end(), std::back_inserter(p3));
    Test::Operations::BoolS r;
    r.resize(p1.size());
    std::reverse_copy(p1.begin(), p1.end(), r.begin());
    return r;
}

Test::Operations::LongS
MyDerivedClassI::opShortIntLongS(const Test::Operations::ShortS& p1,
                                 const Test::Operations::IntS& p2,
                                 const Test::Operations::LongS& p3,
                                 Test::Operations::ShortS& p4,
                                 Test::Operations::IntS& p5,
                                 Test::Operations::LongS& p6,
                                 const Ice::Current&)
{
    p4 = p1;
    p5.resize(p2.size());
    std::reverse_copy(p2.begin(), p2.end(), p5.begin());
    p6 = p3;
    std::copy(p3.begin(), p3.end(), std::back_inserter(p6));
    return p3;
}

Test::Operations::DoubleS
MyDerivedClassI::opFloatDoubleS(const Test::Operations::FloatS& p1,
                                const Test::Operations::DoubleS& p2,
                                Test::Operations::FloatS& p3,
                                Test::Operations::DoubleS& p4,
                                const Ice::Current&)
{
    p3 = p1;
    p4.resize(p2.size());
    std::reverse_copy(p2.begin(), p2.end(), p4.begin());
    Test::Operations::DoubleS r = p2;
    std::copy(p1.begin(), p1.end(), std::back_inserter(r));
    return r;
}

Test::Operations::StringS
MyDerivedClassI::opStringS(const Test::Operations::StringS& p1,
                           const Test::Operations::StringS& p2,
                           Test::Operations::StringS& p3,
                           const Ice::Current&)
{
    p3 = p1;
    std::copy(p2.begin(), p2.end(), std::back_inserter(p3));
    Test::Operations::StringS r;
    r.resize(p1.size());
    std::reverse_copy(p1.begin(), p1.end(), r.begin());
    return r;
}

Test::Operations::ByteSS
MyDerivedClassI::opByteSS(const Test::Operations::ByteSS& p1,
                          const Test::Operations::ByteSS& p2,
                          Test::Operations::ByteSS& p3,
                          const Ice::Current&)
{
    p3.resize(p1.size());
    std::reverse_copy(p1.begin(), p1.end(), p3.begin());
    Test::Operations::ByteSS r = p1;
    std::copy(p2.begin(), p2.end(), std::back_inserter(r));
    return r;
}

Test::Operations::BoolSS
MyDerivedClassI::opBoolSS(const Test::Operations::BoolSS& p1,
                          const Test::Operations::BoolSS& p2,
                          Test::Operations::BoolSS& p3,
                          const Ice::Current&)
{
    p3 = p1;
    std::copy(p2.begin(), p2.end(), std::back_inserter(p3));
    Test::Operations::BoolSS r;
    r.resize(p1.size());
    std::reverse_copy(p1.begin(), p1.end(), r.begin());
    return r;
}

Test::Operations::LongSS
MyDerivedClassI::opShortIntLongSS(const Test::Operations::ShortSS& p1,
                                  const Test::Operations::IntSS& p2,
                                  const Test::Operations::LongSS& p3,
                                  Test::Operations::ShortSS& p4,
                                  Test::Operations::IntSS& p5,
                                  Test::Operations::LongSS& p6,
                                  const Ice::Current&)
{
    p4 = p1;
    p5.resize(p2.size());
    std::reverse_copy(p2.begin(), p2.end(), p5.begin());
    p6 = p3;
    std::copy(p3.begin(), p3.end(), std::back_inserter(p6));
    return p3;
}

Test::Operations::DoubleSS
MyDerivedClassI::opFloatDoubleSS(const Test::Operations::FloatSS& p1,
                                 const Test::Operations::DoubleSS& p2,
                                 Test::Operations::FloatSS& p3,
                                 Test::Operations::DoubleSS& p4,
                                 const Ice::Current&)
{
    p3 = p1;
    p4.resize(p2.size());
    std::reverse_copy(p2.begin(), p2.end(), p4.begin());
    Test::Operations::DoubleSS r = p2;
    std::copy(p2.begin(), p2.end(), std::back_inserter(r));
    return r;
}

Test::Operations::StringSS
MyDerivedClassI::opStringSS(const Test::Operations::StringSS& p1,
                            const Test::Operations::StringSS& p2,
                            Test::Operations::StringSS& p3,
                            const Ice::Current&)
{
    p3 = p1;
    std::copy(p2.begin(), p2.end(), std::back_inserter(p3));
    Test::Operations::StringSS r;
    r.resize(p2.size());
    std::reverse_copy(p2.begin(), p2.end(), r.begin());
    return r;
}

Test::Operations::StringSSS
MyDerivedClassI::opStringSSS(const Test::Operations::StringSSS& p1,
                             const Test::Operations::StringSSS& p2,
                             Test::Operations::StringSSS& p3,
                             const ::Ice::Current&)
{
    p3 = p1;
    std::copy(p2.begin(), p2.end(), std::back_inserter(p3));
    Test::Operations::StringSSS r;
    r.resize(p2.size());
    std::reverse_copy(p2.begin(), p2.end(), r.begin());
    return r;
}

Test::Operations::ByteBoolD
MyDerivedClassI::opByteBoolD(const Test::Operations::ByteBoolD& p1,
                             const Test::Operations::ByteBoolD& p2,
                             Test::Operations::ByteBoolD& p3,
                             const Ice::Current&)
{
    p3 = p1;
    Test::Operations::ByteBoolD r = p1;
    std::set_union(p1.begin(), p1.end(), p2.begin(), p2.end(), std::inserter(r, r.end()));
    return r;
}

Test::Operations::ShortIntD
MyDerivedClassI::opShortIntD(const Test::Operations::ShortIntD& p1,
                             const Test::Operations::ShortIntD& p2,
                             Test::Operations::ShortIntD& p3,
                             const Ice::Current&)
{
    p3 = p1;
    Test::Operations::ShortIntD r = p1;
    std::set_union(p1.begin(), p1.end(), p2.begin(), p2.end(), std::inserter(r, r.end()));
    return r;
}

Test::Operations::LongFloatD
MyDerivedClassI::opLongFloatD(const Test::Operations::LongFloatD& p1,
                              const Test::Operations::LongFloatD& p2,
                              Test::Operations::LongFloatD& p3,
                              const Ice::Current&)
{
    p3 = p1;
    Test::Operations::LongFloatD r = p1;
    std::set_union(p1.begin(), p1.end(), p2.begin(), p2.end(), std::inserter(r, r.end()));
    return r;
}

Test::Operations::StringStringD
MyDerivedClassI::opStringStringD(const Test::Operations::StringStringD& p1,
                                 const Test::Operations::StringStringD& p2,
                                 Test::Operations::StringStringD& p3,
                                 const Ice::Current&)
{
    p3 = p1;
    Test::Operations::StringStringD r = p1;
    std::set_union(p1.begin(), p1.end(), p2.begin(), p2.end(), std::inserter(r, r.end()));
    return r;
}

Test::Operations::StringMyEnumD
MyDerivedClassI::opStringMyEnumD(const Test::Operations::StringMyEnumD& p1,
                                 const Test::Operations::StringMyEnumD& p2,
                                 Test::Operations::StringMyEnumD& p3,
                                 const Ice::Current&)
{
    p3 = p1;
    Test::Operations::StringMyEnumD r = p1;
    std::set_union(p1.begin(), p1.end(), p2.begin(), p2.end(), std::inserter(r, r.end()));
    return r;
}

Test::Operations::MyEnumStringD
MyDerivedClassI::opMyEnumStringD(const Test::Operations::MyEnumStringD& p1,
                                 const Test::Operations::MyEnumStringD& p2,
                                 Test::Operations::MyEnumStringD& p3,
                                 const Ice::Current&)
{
    p3 = p1;
    Test::Operations::MyEnumStringD r = p1;
    std::set_union(p1.begin(), p1.end(), p2.begin(), p2.end(), std::inserter(r, r.end()));
    return r;
}

Test::Operations::MyStructMyEnumD
MyDerivedClassI::opMyStructMyEnumD(const Test::Operations::MyStructMyEnumD& p1,
                                   const Test::Operations::MyStructMyEnumD& p2,
                                   Test::Operations::MyStructMyEnumD& p3,
                                   const Ice::Current&)
{
    p3 = p1;
    Test::Operations::MyStructMyEnumD r = p1;
    std::set_union(p1.begin(), p1.end(), p2.begin(), p2.end(), std::inserter(r, r.end()));
    return r;
}

Test::Operations::IntS
MyDerivedClassI::opIntS(const Test::Operations::IntS& s, const Ice::Current&)
{
    Test::Operations::IntS r;
    std::transform(s.begin(), s.end(), std::back_inserter(r), std::negate<int>());
    return r;
}

void
MyDerivedClassI::opByteSOneway(const Test::Operations::ByteS&, const Ice::Current&)
{
}

Test::Operations::StringStringD
MyDerivedClassI::opContext(const Ice::Current& c)
{
    return c.ctx;
}

void 
MyDerivedClassI::opDoubleMarshaling(Ice::Double p1, const Test::Operations::DoubleS& p2, const Ice::Current&)
{
    Ice::Double d = 1278312346.0 / 13.0;
    test(p1 == d);
    for(unsigned int i = 0; i < p2.size(); ++i)
    {
        test(p2[i] == d);
    }
}

void
MyDerivedClassI::opIdempotent(const Ice::Current& current)
{
    test(current.mode == Ice::Idempotent);
}

void
MyDerivedClassI::opNonmutating(const Ice::Current& current)
{
    test(current.mode == Ice::Nonmutating);
}

void
MyDerivedClassI::opDerived(const Ice::Current&)
{
}
